### DISCUSSION TITLE
Incoming bluetooth data rate

### DISCUSSION THREAD

**Discussion by rbagdazian** [pschatzmann/ESP32-A2DP#366] [Q&A] [answered]
I have written the following function in my ad2f application:

```
auto fftct =0;
auto fftstart=0;
auto fftend=0;
auto fftint=0;

void read_data_stream(const uint8_t *data, uint32_t length) {
  switch(fftct){
    case 0:
      fftstart = millis();
      fftct++;
      break;
    case 1:
      fftend = millis();
      fftct++;
      break;
    case 2:
      fftint = fftend - fftstart;
      fftct++;
      Serial.print("fft interval = ");
      Serial.println(fftint);
      Serial.print("data len =");
      Serial.println(length);
      break;
    default:
      fftct++;
      if(fftct == 50)
        fftct = 0;
      break;
  }
}

void setup() {
   Serial.begin(115200);
   .
   .
   .
   a2dp_sink.set_stream_reader(read_data_stream, false);
   .
   // connect to MyMusic with no automatic reconnect
   a2dp_sink.set_auto_reconnect(false);
   a2dp_sink.start("MyMusic");
}

auto firstTime = 0;
void loop() {
  delay(100); // do nothing
  if(firstTime==0){
    Serial.println("Hello World");
    Serial.print("a2dp-sink sample rate:");
    Serial.println(a2dp_sink.sample_rate());
    Serial.println("--");
    firstTime++;
  }
}


```
When I run the sketch the read_data_stream callback prints the following output:

fft interval = 4
data len =5632

I am trying to understand what this means. The sketch reports a bluetooth sample rate of 44100 which I expect, but the interval between data packets as reported by the read_data_stream callback indicates a packet length of 5632 bytes with an update rate of 4 ms. I am assuming that each sample takes up 4 bytes (2 channels of 16 bit data each). So that the data array consists of 1408 data samples. But this doesn't appear to make sense in that by receiving 1408 samples every four milliseconds this equates to 352000 samples per second instead of the 44100 samples per second I would expect.

I was hoping that someone could enlighten me about where I am going wrong in my understanding.

Thank you!

**Comment by pschatzmann** [comment, marked as answer]
A2DP sends the data as fast as possible: It just gets throttled when you output the data e.g. to I2S.  Since you don't throttle the speed anywere you get this crazy rates!

If you want to do FFT I suggest to use my AudioTools library: here is a [simple fft example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-basic-api/basic-a2dp-fft/basic-a2dp-fft.ino)

**Reply by rbagdazian** [reply]
Yes, I actually used your fft example as my starting point. In this code I just removed the call to fft.write for my testing purposes.
When I changed the initial setup of set_stream_reader with the second parameter as true instead of false as here:
a2dp_sink.set_stream_reader(read_data_stream, true); 
and re-enabled the call to fft.write, the update interval changed to 32 ms which does correspond to what I was expecting.
Thank you!

**Reply by rbagdazian** [reply]
I have a followup question. In your example you show the fft object instantiated as follows: AudioRealFFT fft;
so when the call to fft.write is made does the software only load one channel of the incoming stereo data to the real input to the fft with zero being placed in the imaginary part?  Thank you.

**Reply by pschatzmann** [reply]
If you write 2 channels, they are combined to 1 channel before running fft. 

It depends very much on the implementation how and what information is filled. If you double check the implementation in AudioRealFFT.h you can see that the values are passed as you described above.

**Reply by rbagdazian** [reply]
Ok, thanks. I'll dig into the implementation.
